mne/viz/evoked.py

```diff
@@ -1566,120 +1475,149 @@ def _check_loc_legal(loc, what='your choice', default=1):
-            msg = "Integer colors detected, mapping to rank positions ..."
-            n_colors = len(all_colors)
-            colors_ = {cond: ind for cond, ind in
-                       zip(color_conds, color_indices)}
-
-            def convert_colors(color):
-                return colors_[color]
-        else:
-            for color in all_colors:
-                if not 0 <= color <= 1:
-                    raise ValueError("Values of colors must be all-integer or "
-                                     "floats between 0 and 1, got %s." % color)
-            msg = "Float colors detected, mapping to percentiles ..."
-            n_colors = 101  # percentiles plus 1 if we have 1.0s
-            colors_old = colors.copy()
-
-            def convert_colors(color):
-                return int(colors_old[color] * 100)
-            colors_are_float = True
-        logger.info(msg)
-        the_colors = cmapper(np.linspace(0, 1, n_colors))
-
-        colors = dict()
+        styles = styles.copy()
+        if not set(styles).issubset(tags.union(conditions)):
+            raise ValueError('The keys in "styles" ({}) must '
+                             'match the keys in "evokeds" ({}).'
+                             .format(list(styles), conditions))
+        # make sure all the keys are in there
         for cond in conditions:
-            cond_ = cond.split("/")
-            for color in color_conds:
-                if color in cond_:
-                    colors[cond] = the_colors[convert_colors(color)]
-                    continue
-
-    # categorical colors
+            if cond not in styles:
+                styles[cond] = dict()
+            # deal with matplotlib's synonymous handling of "c" and "color"
+            elif 'c' in styles[cond]:
+                styles[cond]['color'] = styles[cond].pop('c')
+            # transfer styles from partial-matched entries
+            for tag in cond.split('/'):
+                if tag in styles:
+                    styles[cond].update(styles[tag])
+        # remove the (now transferred) partial-matching style entries
+        for key in list(styles):
+            if key not in conditions:
+                del styles[key]
+    # COLORS
+    # make colors a list if it's not defined
+    err_suffix = ''
+    if colors is None:
+        if cmap is None:
+            colors = _get_color_list()
+            err_suffix = ' in the default color cycle'
+        else:
+            colors = list(range(len(conditions)))
+    # convert color list to dict
+    if isinstance(colors, (list, tuple, np.ndarray)):
+        if len(conditions) > len(colors):
+            raise ValueError('Trying to plot {} conditions, but there are '
+                             'only {} colors{}. Please specify colors '
+                             'manually.'
+                             .format(len(conditions), len(colors), err_suffix))
+        colors = dict(zip(conditions, colors))
```

Comment:
Should we check the dimensionality of colors here?

Reply:
we know it's not shorter than `conditions`, so we know that `zip` won't omit any conditions.  I don't think we care if it omits colors (and in fact, when we use the default color cycle it returns a list of 10 colors, so we *want* `zip` to truncate it here).

Reply:
No, I mean, of the individual colors. I.e., that the numeric ones are either one float or three.

Reply:
turns out not to be necessary (I added a pretty thorough test to the gist, which I'll update momentarily)